File: worker/app/vision_runner.py
"""Vision pipeline runner for the VisionJob/Video/AnalysisSession model."""
import asyncio
from datetime import datetime, timezone

from sqlalchemy import select

from app.config import get_settings
from app.db import async_session_maker
from app.models import (
    AnalysisSession, Clip, Event, EventSource, EventType,
    SessionStatus, Video, VisionJob, VisionJobStatus,
)

settings = get_settings()


async def _save_progress(job_id: int, progress: float) -> None:
    """Saves progress in a separate, short-lived DB session (called from the worker thread)."""
    async with async_session_maker() as session:
        job = await session.get(VisionJob, job_id)
        if job and job.status == VisionJobStatus.running:
            job.progress = progress
            await session.commit()


async def run_vision_job(job_id: int) -> None:
    """Runs the video analysis in the background. Updates VisionJob, Event, Clip, Video in the DB."""
    from vision.pipeline import analyze_video

    async with async_session_maker() as session:
        job = await session.get(VisionJob, job_id)
        if job is None:
            return

        video = await session.get(Video, job.video_id)
        if video is None:
            job.status = VisionJobStatus.failed
            job.error = "Video not found"
            job.finished_at = datetime.now(timezone.utc)
            await session.commit()
            return

        job.status = VisionJobStatus.running
        await session.commit()

        video_path = str(settings.uploads_dir / video.filename)

        try:
            # analyze_video is synchronous and CPU-bound (OpenCV + YOLO) — we run it in a
            # separate thread so it doesn't block the event loop (otherwise /status hangs).
            loop = asyncio.get_running_loop()
            last_saved = 0.0  # last saved progress (throttling)

            def on_progress(p: float) -> None:
                nonlocal last_saved
                # save only on a change >= 2% so we don't flood the DB
                if p - last_saved < 0.02 and p < 1.0:
                    return
                last_saved = p
                asyncio.run_coroutine_threadsafe(_save_progress(job_id, p), loop)

            result = await asyncio.to_thread(
                analyze_video,
                video_path,
                yolo_model_path=settings.effective_yolo_model_path,
                frame_stride=settings.frame_stride,
                on_progress=on_progress,
            )

            # save video metadata
            video.duration_seconds = result.duration_seconds
            video.fps = result.fps

            # save events (per-event clips — see below, disabled for now)
            for det in result.events:
                event = Event(
                    analysis_id=video.analysis_id,
                    video_id=video.id,
                    type=EventType(det.type),
                    source=EventSource.auto,
                    timestamp_seconds=det.timestamp_seconds,
                    confidence=det.confidence,
                    label=det.label,
                )
                session.add(event)
                await session.flush()

                # Per-event clip generation — disabled for now.
                # Enable with GENERATE_CLIPS=1; will be useful again later.
                if settings.generate_clips:
                    from vision.clips import extract_clip

                    start = max(0.0, det.timestamp_seconds - settings.clip_pre_seconds)
                    end = det.timestamp_seconds + settings.clip_post_seconds
                    clip_name = f"video{video.id}_event{event.id}.mp4"
                    clip_path = settings.clips_dir / clip_name
                    if extract_clip(video_path, clip_path, start, end):
                        session.add(
                            Clip(
                                event_id=event.id,
                                video_id=video.id,
                                filename=clip_name,
                                start_seconds=start,
                                end_seconds=end,
                            )
                        )

            job.status = VisionJobStatus.done
            job.progress = 1.0
            job.finished_at = datetime.now(timezone.utc)

            # set the session status to "done" if all jobs are finished
            analysis_session = await session.get(AnalysisSession, video.analysis_id)
            if analysis_session:
                all_jobs = (
                    await session.execute(
                        select(VisionJob)
                        .join(Video, VisionJob.video_id == Video.id)
                        .where(Video.analysis_id == video.analysis_id)
                        .where(VisionJob.id != job_id)
                    )
                ).scalars().all()
                all_done = all(j.status == VisionJobStatus.done for j in all_jobs)
                if all_done:
                    analysis_session.status = SessionStatus.done
                    analysis_session.updated_at = datetime.now(timezone.utc)
                    session.add(analysis_session)

            await session.commit()

        except Exception as exc:  # noqa: BLE001
            await session.rollback()
            job = await session.get(VisionJob, job_id)
            if job:
                job.status = VisionJobStatus.failed
                job.error = f"{type(exc).__name__}: {exc}"
                job.finished_at = datetime.now(timezone.utc)
                await session.commit()

File: worker-dotnet/PitchWise.Worker/VisionRunner.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using PitchWise.Api.Config;
using PitchWise.Api.Data;
using PitchWise.Api.Models;
using PitchWise.Vision;

namespace PitchWise.Worker;

/// <summary>
/// Runs one vision job: analyses a video and writes Event/Clip/Video + progress/status.
/// Port of worker/app/vision_runner.py run_vision_job. Uses short-lived DbContext scopes
/// (like the Python short-lived sessions) so progress writes don't hold a long transaction.
/// </summary>
public sealed class VisionRunner
{
    private readonly IServiceScopeFactory _scopeFactory;
    private readonly AppSettings _appSettings;
    private readonly WorkerSettings _worker;
    private readonly IReadOnlyDictionary<int, string> _classNames;

    public VisionRunner(
        IServiceScopeFactory scopeFactory,
        AppSettings appSettings,
        WorkerSettings worker,
        IReadOnlyDictionary<int, string> classNames)
    {
        _scopeFactory = scopeFactory;
        _appSettings = appSettings;
        _worker = worker;
        _classNames = classNames;
    }

    private AppDbContext NewDb(IServiceScope scope) =>
        scope.ServiceProvider.GetRequiredService<AppDbContext>();

    public async Task RunAsync(int jobId, CancellationToken ct)
    {
        // 1. Load job + video, mark running.
        int videoId;
        string videoPath;
        using (IServiceScope scope = _scopeFactory.CreateScope())
        {
            AppDbContext db = NewDb(scope);
            VisionJob? job = await db.VisionJobs.FirstOrDefaultAsync(j => j.Id == jobId, ct);
            if (job is null) return;

            Video? video = await db.Videos.FirstOrDefaultAsync(v => v.Id == job.VideoId, ct);
            if (video is null)
            {
                job.Status = VisionJobStatus.Failed;
                job.Error = "Video not found";
                job.FinishedAt = DateTime.UtcNow;
                await db.SaveChangesAsync(ct);
                return;
            }

            job.Status = VisionJobStatus.Running;
            await db.SaveChangesAsync(ct);

            videoId = video.Id;
            videoPath = Path.Combine(_appSettings.UploadsDir, video.Filename);
        }

        try
        {
            // 2. Single streaming pass: encode a growing HLS VOD with boxes AND detect
            // events in rolling windows. Boxes appear within seconds; the timeline fills
            // in as windows complete; nothing waits for the whole file.
            double lastSaved = 0.0;
            void SaveProgress(double p)
            {
                if (p - lastSaved < 0.02 && p < 1.0) return;
                lastSaved = p;
                _ = SaveProgressAsync(jobId, p);   // fire-and-forget short write
            }

            // HLS lives in a per-video dir under uploads; annotated_filename stores the dir.
            string hlsDirName = $"annotated_v{videoId}";
            string hlsDir = Path.Combine(_appSettings.UploadsDir, hlsDirName);
            if (Directory.Exists(hlsDir))
                try { Directory.Delete(hlsDir, recursive: true); } catch { /* re-analysis: wipe old */ }

            int readyFlagged = 0;
            void OnFirstSegment()
            {
                // Flag playback-ready the moment the first segment lands (once).
                if (Interlocked.Exchange(ref readyFlagged, 1) == 0)
                    _ = SetAnnotatedDirAsync(videoId, hlsDirName);
            }
            void OnEvents(IReadOnlyList<DetectedEvent> evs) =>
                _ = SaveEventsAsync(videoId, evs, videoPath);   // fire-and-forget per window

            // Cooperative cancel: the API sets the job to Failed/cancelled; the worker (a
            // separate process) notices via this DB poll and stops. Cached to keep the poll
            // off the hot path (checked at most ~every 2s inside StreamingAnnotator).
            bool IsCancelled() => IsJobCancelledSync(jobId);

            // ENGINE_DUMP=1 records one engine observation per processed frame next to the HLS
            // output. Replaying it through WorldStateReplay is how the possession and ball-filter
            // thresholds get tuned — without it, every iteration costs a full video pass.
            string? engineDumpPath =
                (Environment.GetEnvironmentVariable("ENGINE_DUMP") ?? "") is "1" or "true"
                    ? Path.Combine(hlsDir, "world_state.jsonl")
                    : null;

            StreamingAnnotator.Result result = await Task.Run(() => StreamingAnnotator.Run(
                videoPath, hlsDir,
                _worker.YoloModelPath, _classNames,
                frameStride: _worker.FrameStride, imgsz: _worker.Imgsz,
                onProgress: SaveProgress,
                onFirstSegment: _worker.RenderAnnotated ? OnFirstSegment : null,
                onEvents: OnEvents,
                executionProvider: _worker.OnnxExecutionProvider,
                deviceId: _worker.OnnxDeviceId,
                isCancelled: IsCancelled,
                reidModelPath: _worker.ReidModelPath,
                // A pitch-keypoint model, when configured, registers each frame to metres per
                // frame (broadcast camera) and switches possession/passes on. Without it the
                // engine stays in normalized coordinates and emits no distance-derived events.
                homography: null,
                engineDumpPath: engineDumpPath,
                pitchModelPath: _worker.PitchModelPath), ct);

            // Persist the per-player time-on-pitch report next to the HLS output (JSON sidecar).
            // Best-effort: a report we can't write shouldn't fail the whole analysis job.
            if (!result.Cancelled && result.TimeOnPitch.Count > 0)
                WriteTimeOnPitchReport(hlsDir, result.TimeOnPitch);

            // 3. Finalise in one scope: duration/fps, ensure annotated dir set, mark done.
            using IServiceScope scope = _scopeFactory.CreateScope();
            AppDbContext db = NewDb(scope);

            Video? video = await db.Videos.FirstOrDefaultAsync(v => v.Id == videoId, ct);
            VisionJob? job = await db.VisionJobs.FirstOrDefaultAsync(j => j.Id == jobId, ct);
            if (video is null || job is null) return;

            // Cancelled: the API already marked the job (Failed/cancelled). Leave its status,
            // just clean up the partial annotated dir and stop.
            if (result.Cancelled)
            {
                try { if (Directory.Exists(hlsDir)) Directory.Delete(hlsDir, recursive: true); } catch { }
                video.AnnotatedFilename = null;
                await db.SaveChangesAsync(ct);
                return;
            }

            video.DurationSeconds = result.DurationSeconds;
            video.Fps = result.Fps;
            if (_worker.RenderAnnotated && result.EncodeOk)
                video.AnnotatedFilename = hlsDirName;

            job.Status = VisionJobStatus.Done;
            job.Progress = 1.0;
            job.FinishedAt = DateTime.UtcNow;

            // 4. Flip the session to "done" when every job across the analysis is finished.
            var siblingStatuses = await db.VisionJobs
                .Join(db.Videos, j => j.VideoId, v => v.Id, (j, v) => new { j.Status, v.AnalysisId })
                .Where(x => x.AnalysisId == video.AnalysisId && x.Status != VisionJobStatus.Done)
                .Select(x => x.Status)
                .ToListAsync(ct);
            // The current job isn't Done in the DB yet (saved below), so exclude it: if no
            // OTHER job is unfinished, this is the last one.
            if (siblingStatuses.Count == 0)
            {
                AnalysisSession? session = await db.AnalysisSessions
                    .FirstOrDefaultAsync(s => s.Id == video.AnalysisId, ct);
                if (session is not null)
                {
                    session.Status = SessionStatus.Done;
                    session.UpdatedAt = DateTime.UtcNow;
                }
            }

            await db.SaveChangesAsync(ct);
        }
        catch (Exception exc)
        {
            using IServiceScope scope = _scopeFactory.CreateScope();
            AppDbContext db = NewDb(scope);
            VisionJob? job = await db.VisionJobs.FirstOrDefaultAsync(j => j.Id == jobId, ct);
            if (job is not null)
            {
                job.Status = VisionJobStatus.Failed;
                job.Error = $"{exc.GetType().Name}: {exc.Message}";
                job.FinishedAt = DateTime.UtcNow;
                await db.SaveChangesAsync(ct);
            }
        }
    }

    // Writes the per-player on-pitch time as a JSON sidecar in the video's HLS dir. Also logs
    // a short summary to the console so it's visible during a manual/verification run.
    private static void WriteTimeOnPitchReport(string hlsDir, IReadOnlyList<PlayerTimeOnPitch> report)
    {
        try
        {
            Directory.CreateDirectory(hlsDir);
            string path = Path.Combine(hlsDir, "time_on_pitch.json");
            string json = System.Text.Json.JsonSerializer.Serialize(
                report, new System.Text.Json.JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(path, json);
            Console.WriteLine($"[Worker] time-on-pitch: {report.Count} players -> {path}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[Worker] time-on-pitch report write skipped: {ex.Message}");
        }
    }

    // Set the annotated HLS dir as soon as the first segment lands, so the frontend can
    // start playing while analysis continues. Best-effort short write.
    private async Task SetAnnotatedDirAsync(int videoId, string hlsDirName)
    {
        try
        {
            using IServiceScope scope = _scopeFactory.CreateScope();
            AppDbContext db = NewDb(scope);
            Video? video = await db.Videos.FirstOrDefaultAsync(v => v.Id == videoId);
            if (video is not null && video.AnnotatedFilename != hlsDirName)
            {
                video.AnnotatedFilename = hlsDirName;
                await db.SaveChangesAsync();
            }
        }
        catch { /* best-effort; finalise step also sets it */ }
    }

    // Persist a window's worth of freshly detected events so the timeline fills in live.
    private async Task SaveEventsAsync(int videoId, IReadOnlyList<DetectedEvent> events, string videoPath)
    {
        if (events.Count == 0) return;
        try
        {
            using IServiceScope scope = _scopeFactory.CreateScope();
            AppDbContext db = NewDb(scope);
            Video? video = await db.Videos.FirstOrDefaultAsync(v => v.Id == videoId);
            if (video is null) return;

            foreach (DetectedEvent det in events)
            {
                var ev = new Event
                {
                    AnalysisId = video.AnalysisId,
                    VideoId = video.Id,
                    Type = EventTypeMap.FromDb(det.Type),
                    Source = EventSource.Auto,
                    TimestampSeconds = det.TimestampSeconds,
                    Confidence = det.Confidence,
                    Label = det.Label,
                };
                db.Events.Add(ev);
                await db.SaveChangesAsync();   // flush to get ev.Id

                if (_worker.GenerateClips)
                {
                    double start = Math.Max(0.0, det.TimestampSeconds - _worker.ClipPreSeconds);
                    double end = det.TimestampSeconds + _worker.ClipPostSeconds;
                    string clipName = $"video{video.Id}_event{ev.Id}.mp4";
                    string clipPath = Path.Combine(_appSettings.ClipsDir, clipName);
                    if (FfmpegTools.ExtractClip(videoPath, clipPath, start, end))
                        db.Clips.Add(new Clip
                        {
                            EventId = ev.Id,
                            VideoId = video.Id,
                            Filename = clipName,
                            StartSeconds = start,
                            EndSeconds = end,
                        });
                }
            }
            await db.SaveChangesAsync();
        }
        catch { /* best-effort; a dropped window just means a few missing events */ }
    }

    // True once the job is no longer Running in the DB — i.e. the API cancelled it (set it
    // to Failed). Synchronous because it's polled from the CPU-bound analysis loop.
    private bool IsJobCancelledSync(int jobId)
    {
        try
        {
            using IServiceScope scope = _scopeFactory.CreateScope();
            AppDbContext db = NewDb(scope);
            VisionJobStatus? status = db.VisionJobs
                .Where(j => j.Id == jobId)
                .Select(j => (VisionJobStatus?)j.Status)
                .FirstOrDefault();
            return status is not null && status != VisionJobStatus.Running;
        }
        catch { return false; }   // transient DB error → keep going, don't cancel spuriously
    }

    private async Task SaveProgressAsync(int jobId, double progress)
    {
        try
        {
            using IServiceScope scope = _scopeFactory.CreateScope();
            AppDbContext db = NewDb(scope);
            VisionJob? job = await db.VisionJobs.FirstOrDefaultAsync(j => j.Id == jobId);
            if (job is not null && job.Status == VisionJobStatus.Running)
            {
                job.Progress = progress;
                await db.SaveChangesAsync();
            }
        }
        catch { /* progress write is best-effort */ }
    }
}
